src/serac/physics/tests/test_solid_functional_statics.cpp

```diff
@@ -0,0 +1,290 @@
+   *
+   * @param X Coordinates of point in reference configuration at which solution is sought
+   * @param u Exact solution evaluated at \p X
+   */
+  void operator()(const mfem::Vector& X, mfem::Vector& u) const
+  {
+    u = 0.0;
+    A.Mult(X, u);
+    u += b;
+  }
+
+  /**
+   * @brief Apply forcing that should produce this exact displacement
+   *
+   * Given the physics module, apply boundary conditions and a source
+   * term that are consistent with the exact solution. This is
+   * independent of the domain. The solution is imposed as an essential
+   * boundary condition on the parts of the boundary identified by \p
+   * essential_boundaries. On the complement of
+   * \p essential_boundaries, the traction corresponding to the exact
+   * solution is applied.
+   *
+   * @tparam p Polynomial degree of the finite element approximation
+   * @tparam Material Type of the material model used in the problem
+   *
+   * @param material Material model used in the problem
+   * @param sf The SolidFunctional module for the problem
+   * @param essential_boundaries Boundary attributes on which essential boundary conditions are desired
+   */
+  template <int p, typename Material>
+  void applyLoads(const Material& material, SolidFunctional<p, dim>& sf, std::set<int> essential_boundaries) const
+  {
+    // essential BCs
+    auto ebc_func = [*this](const auto& X, auto& u){ this->operator()(X, u); };
+    sf.setDisplacementBCs(essential_boundaries, ebc_func);
+
+    // natural BCs
+    typename Material::State state;
+    auto H = make_tensor<dim, dim>([&](int i, int j) { return A(i,j); });
+    tensor<double, dim, dim> tau = material(state, H);
+    auto P = solid_mechanics::KirchhoffToPiola(tau, H);
+    auto traction = [P](auto, auto n0, auto) { return dot(P, n0); };
+    sf.setPiolaTraction(traction);
+  }
+
+ private:
+  mfem::DenseMatrix A; /// Linear part of solution. Equivalently, the displacement gradient
+  mfem::Vector b;      /// Constant part of solution. Rigid mody displacement.
+};
+
+/**
+ * @brief Specify the kinds of boundary condition to apply
+ */
+enum class PatchBoundaryCondition { Essential, Mixed_essential_and_natural };
+
+/**
+ * @brief Get boundary attributes for patch meshes on which to apply essential boundary conditions
+ *
+ * Parameterizes patch tests boundary conditions, as either essential
+ * boundary conditions or partly essential boundary conditions and
+ * partly natural boundary conditions. The return values are specific
+ * to the meshes "patch2d.mesh" and "patch3d.mesh". The particular
+ * portions of the boundary that get essential boundary conditions
+ * are arbitrarily chosen.
+ *
+ * @tparam dim Spatial dimension
+ *
+ * @param b Kind of boundary conditions to apply in the problem
+ * @return std::set<int> Boundary attributes for the essential boundary condition
+ */
+template <int dim>
+std::set<int> essentialBoundaryAttributes(PatchBoundaryCondition bc)
+{
+  if constexpr (dim == 2) {
```

Comment:
It's okay to pass `dim` as a template parameter, but this function doesn't appear to really need compile-time information of the value of `dim`. So, it could be passed just as well as a conventional argument (this would require changing the `if constexpr`s into regular `if`s). Either implementation is acceptable as far as I'm concerned.

Reply:
Good point. I'll leave it for now, just so the `static_assert` on the spatial dimension in the test body doesn't break. Not sure that is the right place to test this condition, but that's probably a subject for another PR.